File: Assignment 5/Python/06. insertion & deletion at a position in a doubly circular ll .py

class node:
    def __init__(self,val):
        self.prev = self
        self.val = val
        self.next = self

    def push_back(self, val):
        nn = node(val)
        if self.next == self and self.prev == self:
            self.next = nn
            self.prev = nn
            nn.prev = self
            nn.next = self
        else:
            ptr = self
            while True:
                ptr = ptr.next
                if ptr.next == self:
                    break
            ptr.next = nn
            nn.prev = ptr
            nn.next = self
            self.prev = nn

    def traversal(self):
        ptr = self
        while True:
            print(ptr.val,end=" ")
            ptr = ptr.next
            if ptr == self:
                break
        print("\n")

def insertion(head,val,pos):
        nn = node(val)
        ptr = head
        count = 2
        while True:
            if count == pos:
                break
            ptr = ptr.next
            count += 1
        nn.next = ptr.next
        nn.prev = ptr
        ptr.next.prev = nn
        ptr.next = nn

def deletion(head,pos):
    ptr = head
    count = 1
    while True:
        if count == pos:
            break
        ptr = ptr.next
        count += 1
    ptr.prev.next = ptr.next
    ptr.next.prev = ptr.prev

head = node(1)
head.push_back(2)
head.push_back(3)
head.push_back(4)
head.traversal()

insertion(head,99,3)
head.traversal()

deletion(head,3)
head.traversal()



# OUTPUT:
# 1 2 3 4 
# 1 2 99 3 4 
# 1 2 3 4

File: Assignment 5/C/06. insertion & deletion at a position in a doubly circular ll.c
#include <stdio.h>
#include <stdlib.h>

typedef struct node{
    struct node* prev;
    int val;
    struct node* next;
}node;

void traversal(node* head){
    node* ptr = head;
    printf("<-->");
    do{
        printf(" %d ",ptr->val);
        ptr = ptr->next;
        if(ptr){
            printf("<-->");
        }
    }while(ptr!=head);
    printf("\n");
}

node* insertion(node* head,int val,int pos){
    node* nn = (node*)malloc(sizeof(node));
    nn->val = val;
    nn->prev = NULL;
    nn->next = NULL;

    node* ptr = head;
    if(pos==1){
        while(ptr->next!=head){
            ptr = ptr->next;
        }
        ptr->next = nn;
        nn->prev = ptr;
        nn->next = head;
        head->prev = nn;
        return nn;
    }
    for(int i=2; ptr->next!=head && i<pos; i++){
        ptr = ptr->next;
    }
    nn->next = ptr->next;
    ptr->next->prev = nn;
    ptr->next = nn;
    nn->prev = ptr;

    return head;
}

node* deletion(node* head,int pos){
    node* ptr = head;
    if(pos==1){
        while(ptr->next!=head){
            ptr = ptr->next;
        }
        ptr->next = ptr->next->next;
        ptr->next->prev = ptr;

        head->prev = NULL;
        head->next = NULL;
        free(head);
        return ptr->next;
    }
    for(int i=2; ptr->next!=NULL && i<pos; i++){
        ptr = ptr->next;
    }
    node* temp = ptr->next;
    
    ptr->next = ptr->next->next;
    ptr->next->prev = ptr; 

    temp->next = NULL;
    temp->prev = NULL;
    free(temp);
    return head;
}

int main(){
    node* head = (node*)malloc(sizeof(node));
    node* snd = (node*)malloc(sizeof(node));
    node* trd = (node*)malloc(sizeof(node));
    node* four = (node*)malloc(sizeof(node));

    head->val = 1;
    snd->val = 2;
    trd->val = 3;
    four->val = 4;

    head->prev = four;
    head->next = snd;

    snd->prev = head;
    snd->next = trd;
    
    trd->prev = snd;
    trd->next = four;

    four->prev = trd;
    four->next = head;

    head = insertion(head,99,4);
    traversal(head);

    head = deletion(head,4);
    traversal(head);

    return 0;
}
